src/stubgen.py

```diff
@@ -81,11 +87,49 @@ class and repeatedly call ``.put()`` to register modules or contents within the
 ]
 # fmt: on
 
+# (name, desired_as_name) -> actual_as_name
+ImportDict = Dict[Tuple[Optional[str], Optional[str]], Optional[str]]
+# package_name -> ((name, desired_as_name) -> actual_as_name)
+PackagesDict = Dict[str, ImportDict]
+
+# (signature_str, doc_str, default_arg_1, default_arg_2, ...)
+# We could type as Tuple[Optional[str], Optional[str], *Tuple[Any, ...]]
+# but this is only valid in python >= 3.11
+NbSignature = Tuple[Optional[str], Optional[str], Any]
+NbGetterSetterSignature = Tuple[str, str]
+
+
+class NbFunction(Protocol):
+    __module__: Literal["nanobind"]
+    __name__: Literal["nb_func", "nb_method"]
+    __nb_signature__: Tuple[NbSignature, ...]
+    __func__: "NbFunction"
```

Comment:
Could you explain why the ``__func__`` and ``__call__`` elements are needed here? Maybe I just don't fully understand how these protocols are used during type-checking.